rs/tests/nns/node_swap_test.rs

```diff
@@ -0,0 +1,211 @@
+use std::time::Duration;
+
+use anyhow::Result;
+use ic_consensus_system_test_utils::rw_message::install_nns_with_customizations_and_check_progress;
+use ic_nervous_system_common_test_keys::{TEST_USER1_KEYPAIR, TEST_USER1_PRINCIPAL};
+use ic_system_test_driver::driver::group::SystemTestGroup;
+use ic_system_test_driver::driver::ic::{InternetComputer, Subnet};
+use ic_system_test_driver::driver::test_env::TestEnv;
+use ic_system_test_driver::driver::test_env_api::{
+    HasTopologySnapshot, IcNodeContainer, IcNodeSnapshot, NnsCustomizations,
+};
+use ic_system_test_driver::retry_with_msg_async;
+use ic_system_test_driver::util::block_on;
+use ic_types::{NodeId, SubnetId};
+use registry_canister::init::RegistryCanisterInitPayloadBuilder;
+use slog::{Logger, info};
+
+const OVERALL_TIMEOUT: Duration = Duration::from_secs(60 * 60);
+
+fn main() -> Result<()> {
+    SystemTestGroup::new()
+        .with_timeout_per_test(OVERALL_TIMEOUT)
+        .with_setup(setup)
+        .add_test(ic_system_test_driver::driver::dsl::TestFunction::new(
+            "node_swaps",
+            test,
+        ))
+        .execute_from_args()
+}
+
+fn setup(env: TestEnv) {
+    // Ensure that the variable is set so that the test panics early
+    // if ic-admin isn't setup correctly.
+    fetch_ic_admin_path();
+
+    let caller = *TEST_USER1_PRINCIPAL;
+    let mut ic = InternetComputer::new()
+        .add_subnet(Subnet::new(ic_registry_subnet_type::SubnetType::System).add_nodes(3))
+        .with_node_operator(caller)
+        .with_node_provider(caller)
+        .with_unassigned_nodes(1);
+
+    ic.setup_and_start(&env)
+        .expect("Failed to start IC under test");
+
+    let snapshot = env.topology_snapshot();
+    let subnet = snapshot.root_subnet();
+
+    let customizations = NnsCustomizations {
+        registry_canister_init_payload: RegistryCanisterInitPayloadBuilder::new()
+            .enable_swapping_feature_globally()
+            .enable_swapping_feature_for_subnet(subnet.subnet_id)
+            .whitelist_swapping_feature_caller(caller)
+            .build(),
+        ..Default::default()
+    };
+
+    install_nns_with_customizations_and_check_progress(env.topology_snapshot(), customizations);
+}
+
+fn fetch_ic_admin_path() -> String {
+    std::env::var("IC_ADMIN_PATH").expect(
+        "IC admin isn't present in the environment variables and is required for this test to run",
+    )
+}
+
+fn test(env: TestEnv) {
+    block_on(test_inner(env))
+}
+
+async fn test_inner(env: TestEnv) {
+    let snapshot = env.topology_snapshot();
```

Comment:
snapshot of what? Many things can be snapshotted. You have dropped the word that's doing most of the story telling!